### PR TITLE
Cleanup and improvements around getting and waiting for documents.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This should be reviewed as https://github.com/dotnet/roslyn/pull/34204/files?w=1

This PR should be reviewed one commit at a time.  It contains the following changes: 

1. Renames `GetOpenDocumentInCurrentContextWithChanges` To `GetDocument`.  GetOpenDocumentInCurrentContextWithChanges was a late addition when we added shared-projects, so we didn't get a good chance to name it properly, but is the method we use all the time.  It's name is full of unnecessary info.  First, 'open' and 'current context' are somewhat pointless.  The starting point is a VS file buffer.  So it's clearly teh Document corresponding to that.  Furthermore, 'WithChanges' is an affection from when it was possible to get a Document that might not be in sync with the snapshot you started with.  This is not a common case, and the immutable nature of VS and Roslyn makes it clear that you want the immutable doc associated with the immutable snap.

2. Rename `GetFullyLoadedOpenDocumentInCurrentContextWithChangesAsync` to `GetFullyLoadedDocument`.  Same reason as above to not have 'open' 'currentcontext' or 'withchanges'.  

3. Made GetFullyLoadedDocument non-async.  All callers are synchronous (by def) since they're blocking some editor operation (with a dialog) for this load to finish.  So the blocking was pushed inward to make all callers simpler.

4. Made GetFullyLoadedDocument operate on a text buffer.  This simplifies things for all callers, and makes more sense from an API perspective as "FullyLoaded" implies mutation will be happening, and that would not affect the results associated with an ITextSnapshot, but would affect an ITextBuffer.

5. Allows one to front-load quick brail-out checks before blocking on waiting for the solution to load.  no point fully loading the project only to immediately bail out a moment later.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @roslyn-ide @dibarbet @jasonmalinowski @dpoeschl.  This PR should be reviewed one commit at a time.

**Comment by sharwell** [comment]
> Made GetFullyLoadedDocument non-async. All callers are synchronous (by def) since they're blocking some editor operation (with a dialog) for this load to finish. So the blocking was pushed inward...

Blocking should always be pushed outward, with the intent to eventually eliminate. This change increases our overall technical debt by moving in an undesirable direction.

File: src/EditorFeatures/Text/Extensions.cs [hunk 1]
```diff
@@ -73,8 +73,18 @@ public static IEnumerable<Document> GetRelatedDocumentsWithChanges(this ITextSna
         /// associated with the buffer if it is linked into multiple projects or is part of a Shared Project. In this case, the <see cref="Workspace"/>
         /// is responsible for keeping track of which of these <see cref="Document"/>s is in the current project context.
         /// </summary>
+        [Obsolete("Use GetDocument instead.", error: true)]
```
Comment (sharwell) [inline src/EditorFeatures/Text/Extensions.cs]: :exclamation: There is no need to set `error: true` here.

1. The method cannot ever be removed; it shipped as a public API
2. The behavior isn't changing, so calling it will continue to have the same effect it always did

:bulb: It would make sense to apply `[EditorBrowsable]` here to prevent the method from appearing in completion.

```suggestion
        [Obsolete("Use GetDocument instead.")]
        [EditorBrowsable(EditorBrowsableState.Never)]
```

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.cs [hunk 2]
```diff
@@ -196,7 +196,7 @@ private static bool IsRenameValid(ConflictResolution conflictResolution, ISymbol
                     if (semanticFactsService.SupportsParameterizedProperties)
                     {
                         otherThingsNamedTheSameExcludeMethodAndParameterizedProperty = otherThingsNamedTheSame
-                            .Where(s=> !s.MatchesKind(SymbolKind.Method, SymbolKind.Property) ||
+                            .Where(s => !s.MatchesKind(SymbolKind.Method, SymbolKind.Property) ||
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.cs]: :grey_exclamation: This file should not be modified in this commit. It is a formatting-only change mixed in with a refactoring that does not need to touch this file.

:memo: The IDE0055 warning is preferable to mixing formatting changes into unrelated refactorings.

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/DeclarationConflictHelpers.cs [hunk 3]
```diff
@@ -18,7 +18,7 @@ public static ImmutableArray<Location> GetMembersWithConflictingSignatures(IMeth
                                             .OfType<IMethodSymbol>()
                                             .Where(m => !m.Equals(renamedMethod) && m.Arity == renamedMethod.Arity);
 
-            return GetConflictLocations(renamedMethod, potentiallyConfictingMethods, isMethod:true,
+            return GetConflictLocations(renamedMethod, potentiallyConfictingMethods, isMethod: true,
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/DeclarationConflictHelpers.cs]: :grey_exclamation: This file should not be modified in this commit. It is a formatting-only change mixed in with a refactoring that does not need to touch this file.

:memo: The IDE0055 warning is preferable to mixing formatting changes into unrelated refactorings.

File: src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs [hunk 4]
```diff
@@ -63,6 +64,17 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
...
+            return GetFullyLoadedDocument(buffer, operationContext, _ => true);
+        }
+
+        /// <summary>
+        /// Get <see cref="Document"/> from <see cref="Text.Extensions.GetDocument(ITextSnapshot)"/>
+        /// once <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync(CancellationToken)"/> returns.
+        /// </summary>
+        /// <param name="shouldLoad">Whether this function should wait for the solution to fully load.</param>
+        public static Document GetFullyLoadedDocument(
+            this ITextBuffer buffer, IUIThreadOperationContext operationContext, Func<Document, bool> shouldLoad)
```
Comment (sharwell) [inline src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs]: :confused: I'm not a big fan of this API. If the checks are moved, I would prefer to see them explicitly placed in the calling code before the call to `GetFullyLoadedDocumentAsync`.

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/DeclarationConflictHelpers.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Rename/ConflictEngine/DeclarationConflictHelpers.cs]: These are errors locally when i compile.  So i fixed so i wouldn't be blocked.  I can remove from the PR though.

File: src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs]: The problem with that is that you now need to get the Doucment in two different ways.  And this leads to a confusing experience in teh code because one Document should only be used for *pre-load-checks*, and one should only be used once the loading has fully happened.  

A problem discovered in the original PR was that it was super easy to do something like:

```c#
var document = ...
//cause fully load
// use 'document'
```

This is try to avoid that problem.

**Comment by CyrusNajmabadi** [comment]
> Blocking should always be pushed outward, with the intent to eventually eliminate. This change increases our overall technical debt by moving in an undesirable direction.

This is not true.  Blocking should be pushed to an appropriate abstraction point.  If you can supply an async commanding model, then it would be fine for this to be async.  HOwever, right now, this only exists in the first place because we have a sync commanding system and it services only that scenario.  So it's fine for it to have this signature.

**Comment by CyrusNajmabadi** [comment]
Closing.

File: src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs (hunk 4 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs]: I still find that situation less confusing than the one here though... Will think on this more

**Comment by CyrusNajmabadi** [comment]
I've closed this out.  Find refs shouldn't even use this system at all. Nor should any async feature. That's because this feature is literally the code that pops up a model dialog forcing you to wait for the solution to load.  Find refs should not do that because it's already async and is a non blocking feature. So if it wants to wait for the solution to fully load, it would not be using this code path.

**Comment by CyrusNajmabadi** [comment]
To be clear:. There should be an *async* way to get the fully loaded solution. But that's not what this function is. This is the function that literally blocks the user behind a model dialog. Being async doesn't end up doing anything.  This is the entry point for that. 

Perhaps this would be clearer of it was an extension on the context object instead and it said 'BlockWithDialogUntilFullyLoaded'?

**Comment by jasonmalinowski** [comment]
> Perhaps this would be clearer of it was an extension on the context object instead and it said 'BlockUIWithDialogUntilFullyLoaded'?

Thtat'd make sense to me. Or honestly I'd just be OK with people doing the two step of start a dialog, and then call the async one and block on it.

(And apologies I'm still behind on some of the design here. Life/family concerns have been taking priority and as a result I haven't had time to look at some of this nearly as much as I want to.)
